Suggested changes for #2

lisa opened the pull request:
Summary of changes for #2:

Makefile:

* Move list of operators into a Makefile variable `$OPERATORS`
* Silence `isclean` output
* Explictly pass `ALLOW_DIRTY_CECKOUT` to "child" `make` calls (eg `make env`)
* Resolve make bash error in `manifests/operators` target
* Use Makefile variable `$OPERATORS` to loop over and fetch source
* Git source fetcher is a little more friendly:
  * Checks to see if the directory exists and git pulls, otherwise clones
  * Resolves minor errors when the directory does exist

scripts/gen_operator_csv.py:

* Fix YAML loading error by changing Loader to use `yaml.BaseLoader` instead of `yaml.FullLoader`

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

lisa committed:
Suggested changes for #2

Summary of changes:

Makefile:

* Move list of operators into a Makefile variable `$OPERATORS`
* Silence `isclean` output
* Explictly pass `ALLOW_DIRTY_CECKOUT` to "child" `make` calls (eg `make env`)
* Resolve make bash error in `manifests/operators` target
* Use Makefile variable `$OPERATORS` to loop over and fetch source
* Git source fetcher is a little more friendly:
** Checks to see if the directory exists and git pulls, otherwise clones
** Resolves minor errors when the directory does exist

scripts/gen_operator_csv.py:

* Fix YAML loading error by changing Loader to use `yaml.BaseLoader` instead of `yaml.FullLoader`

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

rogbas commented on Makefile:
```diff
@@ -48,11 +48,17 @@ CATALOG_VERSION?=$(CHANNEL)-$(BUILD_DATE)-$(CURRENT_COMMIT)
 
 .PHONY: clean
 clean:
+	ifndef SOURCE_DIR
```
This is not working for me, it never identifies SOURCE_DIR, I had to replace with the snippet below to make it work:
```Makefile
ifndef SOURCE_DIR
	$(error SOURCE_DIR needs to be defined or you delete /)
endif
```

rogbas commented on scripts/gen_operator_csv.py:
```diff
@@ -64,7 +64,7 @@
 
     with open('scripts/templates/csv.yaml', 'r') as stream:
-        csv = yaml.load(stream, Loader=yaml.FullLoader)
+        csv = yaml.load(stream, Loader=yaml.BaseLoader)
```
Tested that using `make catalog` and it's adding some garbage to the output file (python!unicode characters) that are making the catalog container crash during startup. The suggestion is going with:
```python
csv = yaml.load(stream)
```
Tested locally and the output is clean.

jewzaam commented on Makefile:
```diff
@@ -48,11 +48,17 @@ CATALOG_VERSION?=$(CHANNEL)-$(BUILD_DATE)-$(CURRENT_COMMIT)
 .PHONY: clean
 clean:
+	ifndef SOURCE_DIR
+	$(error SOURCE_DIR needs to be defined or you delete /)
```
Move this out side of the target and put with the other `ifndef` bits?  It is used in many places, makes sense it has to always be defined.

jewzaam replied:
@lisa looking at this again and it's something we could remove by having `SOURCE_DIR:=operators`, immediate set, instead of current lazy set.

jewzaam commented on Makefile:
```diff
@@ -65,7 +71,7 @@ clean:
 .SILENT: isclean
 isclean:
-	(test "$(ALLOW_DIRTY_CHECKOUT)" != "false" || test 0 -eq $$(git status --porcelain | wc -l)) || (echo "Local git checkout is not clean, commit changes and try again." && exit 1)
+	@(test "$(ALLOW_DIRTY_CHECKOUT)" != "false" || test 0 -eq $$(git status --porcelain | wc -l)) || (echo "Local git checkout is not clean, commit changes and try again." && exit 1)
```
There's `.SILENT: isclean` which should do the same thing but for the whole target.  Is it not working?

lisa replied:
Oh! You're right. Maybe I was seeing the one from the operator? Not sure. I'll remove it

jewzaam replied:
Operator doesn't have it, you're right.  It does @ I think.. I didn't know about `.SILENT` when I did that :D

jewzaam commented on Makefile:
```diff
@@ -102,13 +108,19 @@ manifests/operators: operator-source
+		echo "org = $$org reponame = $$reponame" ; \
+		$(call checkout_operator,$$org,$$reponame) ;\
+		echo ;\
+	done
```
Why have a loop over variable instead of just calling out to the function per repo as it was?  I wouldn't recommend overriding it with CLI vars.  It's just more complicated this way.

lisa replied:
@jewzaam I'm imagining we'll have more than one operator and it's easier to append to the list than duplicate a `$(call ...)` command.

jewzaam replied:
You're editing the Makefile either way.  Felt just another call statement would be very straight forward.  1 line of code per operator, no loop, no var.  Simple.

cblecker replied:
so in my opinion, Makefiles should be pretty static once you get things going. There's lots of dynamic bits that you can do. Iterating over a loop is preferable, I think. It's possible to do things as well as extract the list of operators into a plain text file, and then read that in and iterate over it.

jewzaam replied:
The list of operators will be pretty static once we have them developed and integrated.  I just feel that putting the list into a file that we then have to load an iterate over adds unnecessary complexity in the makefile and adds additional unnecessary places for configuration to be setup.  It's explicit if it's in the target, isn't a var, and isn't in a file.

lisa replied:
I've slept on this and coming back to it I think my main concern about unrolling the loop is the way in which it's unrolled. I believe there's more chance for typos given the Makefile syntax of `$(call checkout_operator,openshift,pagerduty-operator)` compared to appending to a list, `OPERATORS := openshift/dedicated-admin-operator openshift/pagerduty-operator`. We don't often use Makefiles and coming back to remember how to add a new operator to the list is something I want to make as easy and reviewable as possible. For example, imagine you're adding a new operator six months from now and compare:

```diff
diff --git a/Makefile b/Makefile
index 06d1b5d..c9c8067 100644
--- a/Makefile
+++ b/Makefile
@@ -50,7 +50,7 @@ SOURCE_DIR := operators
 
 # List of github.org repositories containing operators
 # This is in the format username/reponame
-OPERATORS := openshift/dedicated-admin-operator
+OPERATORS := openshift/dedicated-admin-operator openshift/pagerduty-operator
 
 .PHONY: default
 default: build
```

Compare to:

```diff
--- Makefile.orig	2019-04-10 08:30:35.000000000 -0400
+++ Makefile	2019-04-10 08:35:56.000000000 -0400
@@ -112,6 +112,7 @@
 .PHONY: operator-source
 operator-source:
 	$(call checkout_operator,openshift,dedicated-admin-operator)
+	$(call checkout_operator,openshift,pagerduty-operator)
 
 .PHONY: catalog
 catalog: operator-source
```

You see a lot of `; \` in the Makefile in other targets, do you need to copy that here? As a reviewer I'm going to wonder the same thing. People reviewing the changes must also be aware that there can't be any whitespace around the `,`, so the natural inclination to add whitespace for readability (like Go demands and Python style encourages) will break it.

I believe the loop (in this PR, for this target) is stable and the path of least resistance is the loop style.

jewzaam commented on checkout-operator.mk:
```diff
@@ -3,7 +3,9 @@
+	if [[ ! -d $(SOURCE_DIR)/$(2) ]]; then \
+		git clone -b master https://github.com/$(1)/$(2).git $(SOURCE_DIR)/$(2) ;\
+	else \
+		cd $(SOURCE_DIR)/$(2) git reset --hard && git pull --force && cd ../.. ;\
```
The reset hard and force pull are actually a HUGE pain if you're doing development.  I had local checkout of the `dedicated-admin-operator` I simply overrode locally in this checkout while I was figuring out what worked with olm.  If this was in there I would need to update the makefile to remove the checkout of source.  Maybe a flag to not have it do this?  Having it as default behavior would be OK.

lisa replied:
That's a good point. I didn't think about doing development work within this checkout. Flag forthcoming.

lisa committed:
Fix syntax error in clean target

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

lisa committed:
Revert `@` before isclean

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

lisa committed:
Loader=BaseLoader results in malformed YAML output

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

lisa committed:
Address Naveen's feedback in checkout-operator.mk

Allow `SKIP_GITREFRESH` to be used to skip doing `git reset` and `git
pull` on the checked out copies of the source code.

Signed-off-by: Lisa Seelye <lseelye@redhat.com>

jewzaam commented on checkout-operator.mk:
```diff
@@ -3,7 +3,13 @@
+		if [[  -z "$${SKIP_GITREFRESH}" ]]; then \
+			cd $(SOURCE_DIR)/$(2) git reset --hard && git pull --force && cd ../.. ;\
+		else \
+			echo "SKIP_GITREFRESH set, skipping git refresh for $(1)$(2)" ;\
```
```suggestion
			echo "SKIP_GITREFRESH set, skipping git refresh for $(1)/$(2)" ;\
```

jewzaam commented on Makefile:
```diff
@@ -48,11 +48,19 @@ CATALOG_VERSION?=$(CHANNEL)-$(BUILD_DATE)-$(CURRENT_COMMIT)
 .PHONY: clean
 clean:
+ifndef SOURCE_DIR
+		$(error SOURCE_DIR needs to be defined or you delete /)
```
```suggestion
	$(error SOURCE_DIR needs to be defined or you delete /)
```

jewzaam committed:
Update checkout-operator.mk

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

lisa committed:
Immediate assignment of operator source dir

jewzaam commented on Makefile:
```diff
@@ -46,7 +46,11 @@ CURRENT_COMMIT=$(shell git rev-parse --short=8 HEAD)
+SOURCE_DIR := operators
+
+# List of github.org repositories containing operators
+# This is in the format username/reponame
```
```suggestion
# This is in the format username/reponame separated by space:  user1/repo1 user2/repo2 user3/repo3
```

jewzaam commented on Makefile:
```diff
@@ -102,13 +106,19 @@ manifests/operators: operator-source
 catalog: operator-source
 	for DIR in $(SOURCE_DIR)/**/; do \
-		eval $$($(MAKE) -C $$DIR env --no-print-directory); \
+		eval $$($(MAKE) ALLOW_DIRTY_CHECKOUT=$(ALLOW_DIRTY_CHECKOUT) -C $$DIR env --no-print-directory); \
```
What's the value in making this and not all the vars used explicit?

```suggestion
		eval $$($(MAKE) -C $$DIR env --no-print-directory); \
```

lisa replied:
What other variables are relevant to `make env`?

lisa replied:
I checked locally and this (and other) vars seem to be inherited so this isn't explicitly needed.

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

jewzaam committed:
Update Makefile

Co-Authored-By: lisa <18159+lisa@users.noreply.github.com>

lisa committed:
this var and others are inherited
